File: search-algorithms/binary-search/go/pkg/binary-search.go
package pkg

// BinarySearch is an iterative implementation of the Binary Search algorithm.
func BinarySearch(haystack []int, needle int) (int, int) {
	var count int
	var low int
	high := len(haystack) - 1

	// While the range to search hasn't shrunk to zero.
	for low <= high {
		count++

		// The mid-point in our range to search.
		mid := (low + high) / 2
		// The value at the midway point in our range.
		found := haystack[mid]

		// We found it!
		if found == needle {
			return mid, count
		}

		if found > needle {
			// The guess is too high, reduce the high index.
			high = mid - 1
			continue
		}

		// The guess is too low, increase the low index.
		low = mid + 1
	}

	return -1, count
}

File: search-algorithms/binary-search/js/src/binary-search.js
/**
 * Binary search implementation using iteration.
 *
 * @param {int[]} haystack - the array to search
 * @param {int} needle - the value to search for
 * @return {Object} - the location of the found value and number of runs
 */
export function binarySearch(haystack, needle) {
	let count = 0;
	let low = 0;
	let high = haystack.length - 1;

	// While the range to search hasn't shrunk to zero.
	while (low <= high) {
		count++;

		// The mid-point in our range to search.
		const mid = Math.floor((low + high) / 2);

		// The guess is too low, increase the low index.
		if (haystack[mid] < needle) {
			low = mid + 1;
			continue;
		}

		// The guess is too high, reduce the high index.
		if (haystack[mid] > needle) {
			high = mid - 1;
			continue;
		}

		// We found it!
		return {foundIdx: mid, runs: count};
	}

	return {foundIdx: -1, runs: count};
}
